examples/src/main/java/io/grpc/examples/manualflowcontrol/ManualFlowControlClient.java

```diff
@@ -0,0 +1,237 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.examples.manualflowcontrol;
+
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.examples.helloworld.HelloReply;
+import io.grpc.examples.helloworld.HelloRequest;
+import io.grpc.examples.helloworld.StreamingGreeterGrpc;
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.ClientResponseObserver;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public class ManualFlowControlClient {
+  public static void main(String[] args) throws InterruptedException {
+    final ExecutorService pool = Executors.newCachedThreadPool();
+    final CountDownLatch done = new CountDownLatch(1);
+
+    // Create a channel and a stub
+    ManagedChannel channel = ManagedChannelBuilder
+        .forAddress("localhost", 50051)
+        .usePlaintext(true)
+        .build();
+    StreamingGreeterGrpc.StreamingGreeterStub stub = StreamingGreeterGrpc.newStub(channel);
+
+    // When using manual flow-control and back-pressure on the client, the ClientResponseObserver handles both
+    // request and response streams.
+    ClientResponseObserver<HelloRequest, HelloReply> clientResponseObserver =
+        new ClientResponseObserver<HelloRequest, HelloReply>() {
+
+          ClientCallStreamObserver<HelloRequest> requestStream;
+
+          @Override
+          public void beforeStart(final ClientCallStreamObserver<HelloRequest> requestStream) {
+            this.requestStream = requestStream;
+            // Set up manual flow control for the response stream. It feels backwards to configure the response
+            // stream's flow control using the request stream's observer, but this is the way it is.
+            requestStream.disableAutoInboundFlowControl();
+
+            // Set up a back-pressure-aware producer for the request stream. The onReadyHandler will be invoked
+            // when the consuming side has enough buffer space to receive more messages.
+            //
+            // Note: the onReadyHandler is invoked by gRPC's internal thread pool. You can't block in this in
+            // method or deadlocks can occur.
+            requestStream.setOnReadyHandler(new Runnable() {
+              // An iterator is used so we can pause and resume iteration of the request data.
+              Iterator<String> iterator = names().iterator();
+
+              @Override
+              public void run() {
+                // Start generating values from where we left off on a non-gRPC thread.
+                pool.execute(new Runnable() {
```

Comment:
Offloading the runnable to another thread pool here is not strictly necessary. gRPC executes user callbacks in donated threads, and you can specify the threadpool in the channel builder with:
https://github.com/grpc/grpc-java/blob/e195c1ab7635f9ae5f2840cbf7247fff14d8513b/core/src/main/java/io/grpc/ManagedChannelBuilder.java#L101

Otherwise, a default thread pool is used. The part of the code that offloads the `Runnable` to the donated tread is: 
https://github.com/grpc/grpc-java/blob/e195c1ab7635f9ae5f2840cbf7247fff14d8513b/core/src/main/java/io/grpc/internal/ClientCallImpl.java#L555